<a href="https://colab.research.google.com/github/kevjavort11/DS_PrimeraEntrega/blob/main/ProyectoDS_ParteI_Ortega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data Science I : Fundamentos para la Ciencia de Datos**

**1RA PRE ENTREGA**

**Fecha: 25/06/2024**

**Estudiante : Ortega Hernandez Kevin Javier**

**Objetivos**



1. Selección uno de los datasets para su evaluacion. (✅)
2. Realizar un abstract (250/500 palabras) del dataset elegido. (✅)
3. Establecer las preguntas e hipótesis de interés sobre el dataset elegido
*   Generar al menos dos preguntas y/o hipotesis. (✅)
4. Realizar gráficos con Matplotlib y con Seaborn
* 3 visualizaciones distintas con Matplotlib: 1 Univariado, 1 Bivariado, 1 libre elección.
* 3 visualizaciones distintas con Seaborn: 1 Univariado, 1 Bivariado, 1 libre elección.
* Generar una interpretación de los gráficos.
* Debe usarse al menos un parametro adicional (grid, hue, etc) que enriquezca
la legibilidad de los charts.





# **Dataset:**

[CTU-IoT-Malware-Capture-48-1](https://docs.google.com/spreadsheets/d/16rvHjh1RfR3YSPoao5vYDXLG3CG1SZYcNRssSlaEe3U)

# **Abstract:**

El dataset *CTU-IoT-Malware-Capture-48-1* consta de veintitrés capturas de diferentes tráficos de red IoT. Estos escenarios se dividen en veinte capturas de red de dispositivos IoT infectados (que tendrán el nombre de la muestra de malware ejecutada en cada escenario) y tres capturas de red de tráfico real de dispositivos IoT.

En cada escenario malicioso se ejecuto una muestra de malware específica en una Raspberry Pi, que utilizó varios protocolos y realizó diferentes acciones.

El tráfico de red capturado para los escenarios benignos se obtuvo capturando el tráfico de red de tres dispositivos IoT diferentes: una lámpara LED inteligente Philips HUE, un asistente personal inteligente para el hogar Amazon Echo y una cerradura inteligente Somfy.

Tanto los escenarios maliciosos como los benignos se ejecutan en un entorno de red controlado con conexión a internet sin restricciones como cualquier otro dispositivo IoT real.

La columna label especifica si es un ataque (1) o si es trafico benigno (0), en la columna multi-label se especifica el tipo de ataque a continuacion una descripcion de los diferentes de las etiquetas en multi-label:

* ***Attack:*** esta etiqueta indica que hubo algún tipo de ataque desde el dispositivo infectado a otro host. Aquí estamos etiquetando como ataque a cualquier flujo que, al analizar su carga útil y comportamiento, intenta aprovechar algún servicio vulnerable. Por ejemplo, un ataque de fuerza bruta a algún inicio de sesión de telnet, una inyección de comandos en el encabezado de una solicitud GET, etc.

* ***Benign:*** esta etiqueta indica que no se encontraron actividades sospechosas o maliciosas en las conexiones.

* ***C&C:*** esta etiqueta indica que el dispositivo infectado estaba conectado a un servidor C&C. Esta actividad se detectó en el análisis de la captura de malware de red porque las conexiones al servidor sospechoso son periódicas o nuestro dispositivo infectado está descargando algunos binarios de él o algunas órdenes similares a IRC o decodificadas van y vienen de él.

* ***FileDownload:*** esta etiqueta indica que se está descargando un archivo a nuestro dispositivo infectado. Esto se detecta filtrando conexiones con bytes de respuesta superiores a 3KB o 5KB, normalmente esto se combina con algún puerto de destino sospechoso conocido o una dirección IP de destino conocida por ser un servidor C&C.

* ***HeartBeat:*** esta etiqueta indica que los paquetes enviados en esta conexión se utilizan para mantener un seguimiento del host infectado por el servidor C&C. Esto se detectó filtrando conexiones con bytes de respuesta inferiores a 1B y con conexiones periódicas similares, normalmente esto se combina con algún puerto de destino sospechoso conocido o una dirección IP de destino conocida por ser un servidor C&C.

* ***PartOfAHorizontalPortScan:*** esta etiqueta indica que las conexiones se utilizan para hacer un escaneo de puertos horizontal para recopilar información para realizar más ataques. Para poner estas etiquetas nos basamos en el patrón en el que las conexiones comparten el mismo puerto, un número similar de bytes transmitidos y múltiples direcciones IP de destino diferentes.

El dataset fue tomado de [Kaggle - Malware Detection in Network Traffic Data](https://www.kaggle.com/datasets/agungpambudi/network-malware-detection-connection-analysis)



# IMPORTACIONES

**Importacion de librerias**

In [ ]:
# Se importaran las librerias necesarias para el desarrollo de las actividades:
# pandas para tratamiento y manipulacion de los datasets
# gdown para la descarga del dataset y poderlo trabajar de forma local
# numpy para las funciones y operaciones matematicas
# matplotlib.pyplot para la visualizacion
# seaborn para visualizacion avanzada

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pandas.errors import ParserError
from google.colab import drive

**Importacion del dataset desde Drive**

In [52]:
drive.flush_and_unmount()
drive.mount('/content/drive')

In [95]:
# Se procede a leer el dataset en formato CSV con pandas
# file_path = '/content/drive/My Drive/CoderHouseDatasets/RT_IOT2022.csv'
# df = pd.read_csv(file_path, sep=',')
file_path = '/content/drive/My Drive/CoderHouseDatasets/CTU-IoT-Malware-Capture-48-1.csv'
df = pd.read_csv(file_path, sep=',')

# Pre-visualizacion del dataset

In [96]:
df.head()

Unnamed: 0  duration  orig_bytes  resp_bytes local_orig local_resp  \
0           0  1.978591         149      119442          -          -   
1           1  2.182247         149      119442          -          -   
2           2  1.675550         152       83118          -          -   
3           3  1.798966         149       88797          -          -   
4           4  1.938142         149      117700          -          -   

   missed_bytes  orig_pkts  orig_ip_bytes  resp_pkts  ...  conn_state_RSTOS0  \
0           0.0      174.0        11698.0      172.0  ...                  0   
1           0.0      172.0        11570.0      170.0  ...                  0   
2           0.0      124.0         8120.0      122.0  ...                  0   
3           0.0      136.0         8578.0      134.0  ...                  0   
4           0.0      170.0        11602.0      170.0  ...                  0   

   conn_state_RSTR  conn_state_RSTRH conn_state_S0  conn_state_S1  \
0                0                 0             0              0   
1                0                 0             0              0   
2                0                 0             0              0   
3                0                 0             0              0   
4                0                 0             0              0   

   conn_state_S2  conn_state_S3  conn_state_SF  conn_state_SH  conn_state_SHR  
0              0              0              1              0               0  
1              0              0              1              0               0  
2              0              0              1              0               0  
3              0              0              1              0               0  
4              0              0              1              0               0  

[5 rows x 37 columns]

In [97]:
df.tail()

Unnamed: 0    duration  orig_bytes  resp_bytes local_orig local_resp  \
40720       40720  133.200167           0           0          -          -   
40721       40721  144.768452           0           0          -          -   
40722       40722  104.514531           0           0          -          -   
40723       40723  233.356672           0           0          -          -   
40724       40724    0.288586           0           0          -          -   

       missed_bytes  orig_pkts  orig_ip_bytes  resp_pkts  ...  \
40720           0.0       16.0          680.0        2.0  ...   
40721           0.0       20.0          880.0        2.0  ...   
40722           0.0       18.0          760.0        2.0  ...   
40723           0.0       16.0          720.0        2.0  ...   
40724           0.0        2.0          120.0        2.0  ...   

       conn_state_RSTOS0  conn_state_RSTR  conn_state_RSTRH conn_state_S0  \
40720                  0                0                 0             0   
40721                  0                0                 0             0   
40722                  0                0                 0             0   
40723                  0                0                 0             0   
40724                  0                0                 0             0   

       conn_state_S1  conn_state_S2  conn_state_S3  conn_state_SF  \
40720              0              1              0              0   
40721              0              1              0              0   
40722              0              1              0              0   
40723              1              0              0              0   
40724              0              0              0              0   

       conn_state_SH  conn_state_SHR  
40720              0               0  
40721              0               0  
40722              0               0  
40723              0               0  
40724              0               0  

[5 rows x 37 columns]

* Se observan datos cualitativos y cuantitativos todas las variables de conn_state_* son cualitativas ya que los valores *0* y *1* solo indican el uso o activacion de esa conexion en el dispositivo IoT, asi como las variables de servicios, y protocolos.

In [79]:
df.shape

(40725, 37)

* Con la ayuda del metodo *shape* se observa la dimensionalidad del dataser con **40725** filas y **37** columnas distribuidas de la siguiente manera:

In [80]:
df.columns

Index(['Unnamed: 0', 'duration', 'orig_bytes', 'resp_bytes', 'local_orig',
       'local_resp', 'missed_bytes', 'orig_pkts', 'orig_ip_bytes', 'resp_pkts',
       'resp_ip_bytes', 'tunnel_parents', 'label', 'multi-label', 'proto_icmp',
       'proto_tcp', 'proto_udp', 'service_-', 'service_dhcp', 'service_dns',
       'service_http', 'service_irc', 'service_ssh', 'service_ssl',
       'conn_state_OTH', 'conn_state_REJ', 'conn_state_RSTO',
       'conn_state_RSTOS0', 'conn_state_RSTR', 'conn_state_RSTRH',
       'conn_state_S0', 'conn_state_S1', 'conn_state_S2', 'conn_state_S3',
       'conn_state_SF', 'conn_state_SH', 'conn_state_SHR'],
      dtype='object')

In [81]:
# El dataset esta depurado y no contiene datos nulos o perdidos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40725 entries, 0 to 40724
Data columns (total 37 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         40725 non-null  int64  
 1   duration           40725 non-null  float64
 2   orig_bytes         40725 non-null  int64  
 3   resp_bytes         40725 non-null  int64  
 4   local_orig         40725 non-null  object 
 5   local_resp         40725 non-null  object 
 6   missed_bytes       40725 non-null  float64
 7   orig_pkts          40725 non-null  float64
 8   orig_ip_bytes      40725 non-null  float64
 9   resp_pkts          40725 non-null  float64
 10  resp_ip_bytes      40725 non-null  float64
 11  tunnel_parents     40725 non-null  int64  
 12  label              40725 non-null  int64  
 13  multi-label        40725 non-null  object 
 14  proto_icmp         40725 non-null  int64  
 15  proto_tcp          40725 non-null  int64  
 16  proto_udp          407

In [102]:
# Se procede a analizar a fondo las columnas para identificar realmente el tipo
# de variables que contienen, determinando cuales son los valores unicos de cada
# columna podremos hacernos una idea de que tipo de variable contiene

unique_values = {col: df[col].unique() for col in df.columns}
for col, values in unique_values.items():
  print(f"Valores unicos en la columna {col} : {values}")
  # print(f"{col} ")

Valores unicos en la columna Unnamed: 0 : [    0     1     2 ... 40722 40723 40724]
Valores unicos en la columna duration : [  1.978591   2.182247   1.67555  ... 104.514531 233.356672   0.288586]
Valores unicos en la columna orig_bytes : [ 149  152    0  148  147   96  120   16   27   56  136  240  104  415
   22  992  280  576   24   19  421  320   12 1246   15   88   25  360
   31  900  112 1237  424   80  272    1    4   21 2364  160   13   28
   29   30   23  480]
Valores unicos en la columna resp_bytes : [119442  83118  88797 117700      0  75854  88857  74766  89651 103261
 102326     96    856     81    118     61     72     39  25840    576
     37     36     38    114     83     56    690     25    438    302
    513     86    228     91     62    590    240     94    100    303
     87     27    496     74    166    101    154     45    593    109
     33    610     54     92     47     30     34    119     97     79
   1884     15     77     32    151    135    171     95   

* Con base en lo anterior podemos determinar que las siguientes columnas contienen valores categoricos ya que los numeros *0* y *1* representan el uso o no de dicho protocolo, servicio o conexion
  * proto_icmp,
proto_tcp,
proto_udp,
service_-,
service_dhcp,
service_dns,
service_http,
service_irc,
service_ssh,
service_ssl,
conn_state_OTH,
conn_state_REJ,
conn_state_RSTO,
conn_state_RSTOS0,
conn_state_RSTR,
conn_state_RSTRH,
conn_state_S0,
conn_state_S1,
conn_state_S2,
conn_state_S3,
conn_state_SF,
conn_state_SH,
conn_state_SHR,
label,
tunnel_parents

* Mientras que las siguentes columnas contienen valores numericos
  * duration,
orig_bytes,
resp_bytes,
local_orig,
local_resp,
missed_bytes,
orig_pkts,
orig_ip_bytes,
resp_pkts,
resp_ip_bytes

* De lo anterior tambien se puede apreciar que las columnas local_orig, local_resp no aportan ningun tipo de informacion relevante, por lo que se procede a eliminarlas

In [103]:
df = df.drop(columns=['local_orig', 'local_resp'])

# Display the first few rows of the updated dataframe to confirm the columns are dropped
df.head()

Unnamed: 0  duration  orig_bytes  resp_bytes  missed_bytes  orig_pkts  \
0           0  1.978591         149      119442           0.0      174.0   
1           1  2.182247         149      119442           0.0      172.0   
2           2  1.675550         152       83118           0.0      124.0   
3           3  1.798966         149       88797           0.0      136.0   
4           4  1.938142         149      117700           0.0      170.0   

   orig_ip_bytes  resp_pkts  resp_ip_bytes  tunnel_parents  ...  \
0        11698.0      172.0       247844.0               0  ...   
1        11570.0      170.0       247740.0               0  ...   
2         8120.0      122.0       172596.0               0  ...   
3         8578.0      134.0       193266.0               0  ...   
4        11602.0      170.0       244256.0               0  ...   

   conn_state_RSTOS0 conn_state_RSTR  conn_state_RSTRH  conn_state_S0  \
0                  0               0                 0              0   
1                  0               0                 0              0   
2                  0               0                 0              0   
3                  0               0                 0              0   
4                  0               0                 0              0   

   conn_state_S1  conn_state_S2  conn_state_S3  conn_state_SF  conn_state_SH  \
0              0              0              0              1              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              0              0              0              1              0   
4              0              0              0              1              0   

   conn_state_SHR  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 35 columns]

In [104]:
df.columns

Index(['Unnamed: 0', 'duration', 'orig_bytes', 'resp_bytes', 'missed_bytes',
       'orig_pkts', 'orig_ip_bytes', 'resp_pkts', 'resp_ip_bytes',
       'tunnel_parents', 'label', 'multi-label', 'proto_icmp', 'proto_tcp',
       'proto_udp', 'service_-', 'service_dhcp', 'service_dns', 'service_http',
       'service_irc', 'service_ssh', 'service_ssl', 'conn_state_OTH',
       'conn_state_REJ', 'conn_state_RSTO', 'conn_state_RSTOS0',
       'conn_state_RSTR', 'conn_state_RSTRH', 'conn_state_S0', 'conn_state_S1',
       'conn_state_S2', 'conn_state_S3', 'conn_state_SF', 'conn_state_SH',
       'conn_state_SHR'],
      dtype='object')

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40725 entries, 0 to 40724
Data columns (total 35 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         40725 non-null  int64  
 1   duration           40725 non-null  float64
 2   orig_bytes         40725 non-null  int64  
 3   resp_bytes         40725 non-null  int64  
 4   missed_bytes       40725 non-null  float64
 5   orig_pkts          40725 non-null  float64
 6   orig_ip_bytes      40725 non-null  float64
 7   resp_pkts          40725 non-null  float64
 8   resp_ip_bytes      40725 non-null  float64
 9   tunnel_parents     40725 non-null  int64  
 10  label              40725 non-null  int64  
 11  multi-label        40725 non-null  object 
 12  proto_icmp         40725 non-null  int64  
 13  proto_tcp          40725 non-null  int64  
 14  proto_udp          40725 non-null  int64  
 15  service_-          40725 non-null  int64  
 16  service_dhcp       407

In [94]:
df.describe()

Unnamed: 0      duration    orig_bytes     resp_bytes  missed_bytes  \
count  40725.000000  40725.000000  40725.000000   40725.000000       40725.0   
mean   20362.000000      3.850978      1.937287      28.044371           0.0   
std    11756.439193    112.729969     24.051290    1615.159047           0.0   
min        0.000000      0.000000      0.000000       0.000000           0.0   
25%    10181.000000      3.094776      0.000000       0.000000           0.0   
50%    20362.000000      3.117713      0.000000       0.000000           0.0   
75%    30543.000000      3.140472      0.000000       0.000000           0.0   
max    40724.000000  22673.812724   2364.000000  119442.000000           0.0   

          orig_pkts  orig_ip_bytes     resp_pkts  resp_ip_bytes  \
count  40725.000000   40725.000000  40725.000000   40725.000000   
mean       5.929282     354.824948      0.388901      73.201621   
std       12.181439     666.626229     12.303606    3438.077790   
min        1.000000      60.000000      0.000000       0.000000   
25%        6.000000     360.000000      0.000000       0.000000   
50%        6.000000     360.000000      0.000000       0.000000   
75%        6.000000     360.000000      0.000000       0.000000   
max     2326.000000  126522.000000   2362.000000  247844.000000   

       tunnel_parents  ...  conn_state_RSTOS0  conn_state_RSTR  \
count         40725.0  ...       40725.000000     40725.000000   
mean              0.0  ...           0.000516         0.001621   
std               0.0  ...           0.022702         0.040225   
min               0.0  ...           0.000000         0.000000   
25%               0.0  ...           0.000000         0.000000   
50%               0.0  ...           0.000000         0.000000   
75%               0.0  ...           0.000000         0.000000   
max               0.0  ...           1.000000         1.000000   

       conn_state_RSTRH  conn_state_S0  conn_state_S1  conn_state_S2  \
count           40725.0   40725.000000   40725.000000   40725.000000   
mean                0.0       0.902124       0.000614       0.000614   
std                 0.0       0.297151       0.024769       0.024769   
min                 0.0       0.000000       0.000000       0.000000   
25%                 0.0       1.000000       0.000000       0.000000   
50%                 0.0       1.000000       0.000000       0.000000   
75%                 0.0       1.000000       0.000000       0.000000   
max                 0.0       1.000000       1.000000       1.000000   

       conn_state_S3  conn_state_SF  conn_state_SH  conn_state_SHR  
count        40725.0   40725.000000   40725.000000         40725.0  
mean             0.0       0.009576       0.000295             0.0  
std              0.0       0.097391       0.017163             0.0  
min              0.0       0.000000       0.000000             0.0  
25%              0.0       0.000000       0.000000             0.0  
50%              0.0       0.000000       0.000000             0.0  
75%              0.0       0.000000       0.000000             0.0  
max              0.0       1.000000       1.000000             0.0  

[8 rows x 34 columns]

['-']
['-']



# **Preguntas Hipotesis**

1. ¿La duración de las conexiones de red difiere significativamente entre los dispositivos IoT benignos y los infectados por malware?

2. ¿Existe una diferencia significativa en el número de bytes transmitidos (tanto originales como de respuesta) entre los dispositivos IoT benignos y los infectados por malware?

3. ¿Los dispositivos IoT infectados por malware tienen más bytes perdidos en comparación con los dispositivos benignos?

4. ¿El número de paquetes (tanto originales como de respuesta) es significativamente diferente entre los dispositivos IoT benignos y los infectados por malware?

5. ¿Los diferentes estados de conexión (por ejemplo, SF, S0, S1, etc.) ocurren con diferentes frecuencias en los dispositivos IoT benignos frente a los infectados por malware?